lib/winapi/timezoneapi.c

```diff
@@ -0,0 +1,197 @@
+    LONG timeNow, dayLightStart, dayLightEnd;
+    XBOX_TZ_STRUCT dayLightStartDate, dayLightEndDate;
+    CHAR dayLightName[4], standardName[4];
+    NTSTATUS queryStatus;
+    TIME_FIELDS dateNow;
+    LARGE_INTEGER kTime;
+    ULONG type;
+
+    RtlZeroMemory(lpTimeZoneInformation, sizeof(TIME_ZONE_INFORMATION));
+
+    queryStatus = ExQueryNonVolatileSetting(XC_TIMEZONE_BIAS, &type, &timeZoneBias, sizeof(timeZoneBias), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    lpTimeZoneInformation->Bias = timeZoneBias;
+
+    // Check if daylight savings time (DST) adjustment is enabled
+    queryStatus = ExQueryNonVolatileSetting(XC_MISC, &type, &dayLightEnFlag, sizeof(dayLightEnFlag), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    // DST adjustment is disabled
+    if (dayLightEnFlag & 0x2) {
+        return TIME_ZONE_ID_UNKNOWN;
+    }
+
+    // Query all DST info from EEPROM
+    // FIXME One large EEPROM access could be better
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_DLT_DATE, &type, &dayLightStartDate, sizeof(dayLightStartDate), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_DATE, &type, &dayLightEndDate, sizeof(dayLightEndDate), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_DLT_BIAS, &type, &dayLightBias, sizeof(dayLightBias), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_BIAS, &type, &standardBias, sizeof(standardBias), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_DLT_NAME, &type, dayLightName, sizeof(dayLightName), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_NAME, &type, standardName, sizeof(standardName), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    if (dayLightEndDate.Month > 12 || dayLightStartDate.Month > 12 ||
+        dayLightEndDate.DayOfWeek > 6 || dayLightStartDate.DayOfWeek > 6 ||
+        dayLightEndDate.Hour > 23 || dayLightStartDate.Hour > 23) {
+        return TIME_ZONE_ID_INVALID;
+    }
```

Comment:
This would also have to check if `Month < 1`?

nit: You might also want to move this into a helper function to check if a date is valid.

Reply:
When month == 0 does not return TIME_ZONE_ID_INVALID, it returns TIME_ZONE_ID_UNKNOWN.

So the check for month ==0 is done further down in the code. On xbox it also appears to still attempt to copy the standardName and dayLightName even if month == 0 so my code does that too.

Reply:
moved it to a helper function